### DISCUSSION TITLE
Research papers

### DISCUSSION THREAD

**Discussion by alexmojaki** [aroberge/friendly#197] [Ideas]
I was browsing through research papers and found a few I thought you'd like:

- [Frequency Distribution of Error Messages](https://arxiv.org/pdf/1509.07238.pdf) shows how common various Python errors are in their online learning environment [CS Circles](https://cscircles.cemc.uwaterloo.ca/). The data is [here](http://daveagp.github.io/errors/python.txt).
- [Python3.6 Natural Syntax Errors and Corrections](https://figshare.com/articles/dataset/Python3_6_Natural_Syntax_Errors_and_Corrections/8244686/1) is a dataset of erroneous code, mostly syntax errors, from StackOverflow. You can probably automatically find plenty of cases not handled by friendly. Here's some code for working with it:

```python
import json
from pathlib import Path

path = Path("/home/alex/Downloads/8244686")

parse_errors = json.loads((path / "parse_errors.json").read_text())
parse_fixes = json.loads((path / "parse_fixes.json").read_text())
pairs = list(zip(parse_errors.values(), parse_fixes.values()))
for error, fix in pairs[:10]:
    error = error["content"]
    fix = fix["content"]
    print(error)
    print("---")
    print(fix)
    print("---")
    print("---")
```

- [The Error Behind The Message: Finding the Cause of Error Messages in Python](https://tobiaskohn.ch/files/TKohn_ErrorBehingMessage.pdf) generally talks about error messages and beginners. The same author also wrote [Tell Me What's Wrong: A Python IDE with Error Messages](https://dl.acm.org/doi/10.1145/3328778.3366920) more recently which looks more relevant but I can't access the full text. I think the relevant source code is [here](https://github.com/Tobias-Kohn/TigerPython-Parser) which is a Python parser written in Scala that supposedly gives better error messages.
- [A System for Classifying and Clarifying Python Syntax Errors for Educational Purposes](https://dspace.mit.edu/bitstream/handle/1721.1/119750/1078691087-MIT.pdf?sequence=1&isAllowed=y). An interesting idea here is using parso as a way of locating syntax errors. For example:

```python
import parso

code = """
x = 4
if x == 4
    print ('cat ')
"""

result = parso.parse(code)
print(result.children[1])  # <PythonErrorNode: if x == 4@3,0>
print(result.children[1].get_code())  # if x == 4
```

**Comment by aroberge** [comment]
This is fantastic, thank you. 

I was already familiar with Tobias Kohn work (and have noted some of his examples I found in one of his repository). I didn't know that CsCircle had that resource; I did some translation work for them a few years ago.

This should keep me busy for a while! :-)

**Comment by aroberge** [comment]
The IDE in question is TygerJython http://jython.tobiaskohn.ch/index.html.  It also has a web based version https://webtigerjython.ethz.ch/

**Comment by aroberge** [comment]
I turned the list of errors from Waterloo's Cs Circles into an issue: https://github.com/aroberge/friendly-traceback/issues/198

**Comment by alexmojaki** [comment]
@pablogsal

**Reply by pablogsal** [reply]
Thanks for pointing out this, I will surely check all these once I have some available time. Thanks a lot 🤘
